### PR TITLE
Controller typing

### PR DESCRIPTION
**Pull request by chrisFrazier77**

### CODE REVIEW

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts [hunk 1]
```diff
@@ -313,10 +314,10 @@ export class AutocompleteController extends AbstractController {
 
 				this.store.state.input = value;
 
-				if (this.config.settings.syncInputs) {
+				if (this.config?.settings?.syncInputs) {
 					const inputs = document.querySelectorAll(this.config.selector);
-					inputs.forEach((input: HTMLInputElement) => {
-						input.value = value;
+					inputs.forEach((input) => {
```
Comment (korgon) [inline packages/snap-controller/src/Autocomplete/AutocompleteController.ts:319]: This change is interesting - why move the typing out of the function?

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts [hunk 2]
```diff
@@ -338,7 +339,7 @@ export class AutocompleteController extends AbstractController {
 					}, INPUT_DELAY);
 				}
 			},
-			timeoutDelay: undefined,
+			timeoutDelay: undefined as undefined | ReturnType<typeof setTimeout>,
```
Comment (korgon) [inline packages/snap-controller/src/Autocomplete/AutocompleteController.ts:342]: This is a bit odd `undefined as undefined` ?

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts [hunk 3]
```diff
@@ -353,13 +354,13 @@ export class AutocompleteController extends AbstractController {
...
-		const inputs = document.querySelectorAll(`input[${INPUT_ATTRIBUTE}]`);
-		inputs?.forEach((input: HTMLInputElement) => {
+		const inputs: NodeListOf<HTMLInputElement> = document.querySelectorAll(`input[${INPUT_ATTRIBUTE}]`);
+		inputs?.forEach((input) => {
 			input.removeEventListener('keyup', this.handlers.input.keyUp);
 			input.removeEventListener('keydown', this.handlers.input.enterKey);
 			input.removeEventListener('keydown', this.handlers.input.escKey);
 			input.removeEventListener('focus', this.handlers.input.focus);
-			input.form?.removeEventListener('submit', this.handlers.input.formSubmit);
+			input.form?.removeEventListener('submit', this.handlers.input.formSubmit as unknown as EventListener);
```
Comment (korgon) [inline packages/snap-controller/src/Autocomplete/AutocompleteController.ts:363]: Is this necessary?

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts [hunk 4]
```diff
@@ -416,7 +417,7 @@ export class AutocompleteController extends AbstractController {
 			}
 		});
 
-		if (this.config.settings?.trending?.limit > 0 && !this.store.trending?.length) {
+		if (this.config.settings?.trending?.limit && this.config.settings?.trending?.limit > 0 && !this.store.trending?.length) {
```
Comment (korgon) [inline packages/snap-controller/src/Autocomplete/AutocompleteController.ts:420]: Shouldn't the optional chaining handle this?

File: packages/snap-controller/src/Finder/FinderController.test.ts [hunk 5]
```diff
@@ -38,7 +38,7 @@ const configs = [
 describe('Finder Controller', () => {
 	configs.forEach((baseConfig) => {
 		const isHierarchy = 'levels' in baseConfig.fields[0];
-		let urlManager, services, config;
+		let urlManager: UrlManager, services: any, config: any;
```
Comment (korgon) [inline packages/snap-controller/src/Finder/FinderController.test.ts:41]: Can `services` and `config` have types other than `any`?

File: packages/snap-controller/src/Finder/FinderController.test.ts [hunk 6]
```diff
@@ -69,7 +69,7 @@ describe('Finder Controller', () => {
 				await controller.search();
 
 				const firstSelection = controller.store.selections[0];
-				const valueToSelect = firstSelection.values.filter((value) => value.count > 10)[0].value;
+				const valueToSelect = firstSelection.values.filter((value: any) => value.count > 10)[0].value;
```
Comment (korgon) [inline packages/snap-controller/src/Finder/FinderController.test.ts:72]: Can we use actual type here (and in other places where `any` is utilized)?

File: packages/snap-controller/src/Finder/FinderController.ts [hunk 7]
```diff
@@ -35,6 +35,9 @@ export class FinderController extends AbstractController {
 	) {
 		super(config, { client, store, urlManager, eventManager, profiler, logger, tracker }, context);
 
+		this.store = store as FinderStore;
+		this.config = config;
```
Comment (korgon) [inline packages/snap-controller/src/Finder/FinderController.ts:39]: The abstract controller already does this.

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts [hunk 8]
```diff
@@ -57,13 +57,14 @@ export class AutocompleteController extends AbstractController {
 	) {
 		super(config, { client, store, urlManager, eventManager, profiler, logger, tracker }, context);
 
+		this.store = store as AutocompleteStore;
 		// deep merge config with defaults
-		this.config = deepmerge(defaultConfig, this.config);
+		this.config = deepmerge(defaultConfig, config);
```
Comment (korgon) [inline packages/snap-controller/src/Autocomplete/AutocompleteController.ts:62]: Why change this to `config` vs. `this.config` ?

File: packages/snap-controller/src/Finder/FinderController.ts [hunk 9]
```diff
@@ -97,7 +100,7 @@ export class FinderController extends AbstractController {
 			await this.eventManager.fire('beforeFind', {
 				controller: this,
 			});
-		} catch (err) {
+		} catch (err: any) {
```
Comment (korgon) [inline packages/snap-controller/src/Finder/FinderController.ts:103]: Can errors be typed?

File: packages/snap-controller/src/Recommendation/RecommendationController.test.ts [hunk 10]
```diff
@@ -246,9 +246,9 @@ describe('Recommendation Controller', () => {
 		expect(trackfn).toHaveBeenCalledTimes(controller.store.results.length + 1);
 		expect(productTrackfn).toHaveBeenCalledTimes(controller.store.results.length);
 
-		expect(Object.keys(controller.events.product)).toHaveLength(controller.store.results.length);
-		expect(controller.events.product[controller.store.results[0].id]).toBeDefined();
-		expect(controller.events.product[controller.store.results[0].id].render).toBeDefined();
+		expect(Object.keys(controller.events.product || {})).toHaveLength(controller.store.results.length);
+		expect(controller.events.product![controller.store.results[0].id as keyof BeaconEvent]).toBeDefined();
```
Comment (korgon) [inline packages/snap-controller/src/Recommendation/RecommendationController.test.ts:250]: This type seems odd to me. Why `BeaconEvent` ?

File: packages/snap-controller/src/Recommendation/RecommendationController.ts [hunk 11]
```diff
@@ -126,17 +147,17 @@ export class RecommendationController extends AbstractController {
 					pid: this.events.impression.id,
 				};
 
-				this.events.product[result.id] = this.events.product[result.id] || {};
-				const event = (this.events.product[result.id].impression = this.tracker.track.event(payload));
+				this.events.product![result.id as keyof BeaconEvent] = this.events.product![result.id as keyof BeaconEvent] || {};
```
Comment (korgon) [inline packages/snap-controller/src/Recommendation/RecommendationController.ts:150]: This type seems odd to me. Why `BeaconEvent`?

**Review by korgon — CHANGES_REQUESTED** [review]

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts [hunk 12]
```diff
@@ -255,9 +256,9 @@ export class AutocompleteController extends AbstractController {
 					this.setFocused(e.target as HTMLInputElement);
 				});
 			},
-			formSubmit: async (e): Promise<void> => {
+			formSubmit: async (e: React.FormEvent<HTMLInputElement>): Promise<void> => {
 				const form = e.target;
```
[the remark this answers is not in the document]
Comment (chrisFrazier77) [reply packages/snap-controller/src/Autocomplete/AutocompleteController.ts:260]: not worth it?... - Kevin

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts (hunk 1 above)
Comment (chrisFrazier77) [reply packages/snap-controller/src/Autocomplete/AutocompleteController.ts:319]: needed typing

File: packages/snap-controller/src/Autocomplete/AutocompleteController.ts (hunk 3 above)
Comment (chrisFrazier77) [reply packages/snap-controller/src/Autocomplete/AutocompleteController.ts:363]: YEP

File: packages/snap-controller/src/types.ts [hunk 13]
```diff
@@ -54,7 +57,7 @@ export enum ControllerTypes {
 
 export type ControllerServices = {
 	client: Client;
-	store: AbstractStore;
+	store: SearchStore | AutocompleteStore | FinderStore | RecommendationStore;
```
[the remark this answers is not in the document]
Comment (chrisFrazier77) [reply packages/snap-controller/src/types.ts:60]: createcontrollers use this type and need it to match this.

**Review by korgon — APPROVED** [review]
Think we are there... merge away!

**Pull request merged**